lib/src/main/java/com/dynatrace/file/util/PollBasedFilePoller.java

```diff
@@ -59,21 +74,86 @@ public boolean fileContentsUpdated() {
     return changedSinceLastInquiry.getAndSet(false);
   }
 
+  private synchronized void initialPoll() {
+    try {
+      prevChecksum = getFileChecksum();
+      prevModMillis =
+          Files.getLastModifiedTime(absoluteFilePath, LinkOption.NOFOLLOW_LINKS).toMillis();
+    } catch (IOException e) {
+      logger.warning(
+          () ->
+              String.format(
+                  "Failed to read file %s, stopping polling. Error: %s", absoluteFilePath, e));
+      worker.cancel(true);
+    }
+  }
+
   private synchronized void poll() {
     try {
-      FileTime lastModifiedAt = Files.getLastModifiedTime(absoluteFilePath);
+      // check if the filetime has changed.
+      long lastModifiedMillis = Files.getLastModifiedTime(absoluteFilePath).toMillis();
+      if (lastModifiedMillis > prevModMillis) {
+        prevModMillis = lastModifiedMillis;
+        changedSinceLastInquiry.set(true);
+        return;
+      }
 
-      if (lastModifiedAt.compareTo(prevModifiedAt) >= 0) {
-        if (prevModifiedAt.compareTo(ZERO_FILE_TIME) != 0) {
-          changedSinceLastInquiry.set(true);
-        }
+      // if the filetime hasn't changed, check if the hash has changed.
+      String fileChecksum = getFileChecksum();
+      if (!fileChecksum.equals(prevChecksum)) {
+        prevChecksum = fileChecksum;
+        prevModMillis = lastModifiedMillis;
+        changedSinceLastInquiry.set(true);
       }
-      prevModifiedAt = lastModifiedAt;
     } catch (IOException e) {
       // One possible reason for this is that no read permissions exist on the file, in
       // which case the user should (try to) read the file anyway and handle any errors then.
       changedSinceLastInquiry.set(true);
-      logger.warning(() -> String.format("Failed to read file %s; Error: %s", absoluteFilePath, e));
+      logger.warning(
+          () ->
+              String.format(
+                  "Failed to read file %s, stopping polling. Error: %s", absoluteFilePath, e));
+      worker.cancel(true);
+    }
+  }
+
+  private synchronized String getFileChecksum() throws IOException {
+    // Get file input stream for reading the file content
+    try (FileInputStream fis = new FileInputStream(absoluteFilePath.toFile())) {
+
+      // Create byte array to read data in chunks
+      byte[] byteArray = new byte[1024];
+      int bytesCount;
+
+      // Read file data and update in message digest
+      while ((bytesCount = fis.read(byteArray)) != -1) {
+        md5.update(byteArray, 0, bytesCount);
+      }
+    }
+
+    // Get the hash's bytes
+    byte[] bytes = md5.digest();
+
+    // This bytes[] has bytes in decimal format, convert it to hexadecimal format
+    StringBuilder sb = new StringBuilder();
+    for (byte aByte : bytes) {
+      sb.append(Integer.toString((aByte & 0xff) + 0x100, 16).substring(1));
+    }
+
+    // return complete hash
+    return sb.toString();
```

Comment:
You could also just return the bytes and compare them with `Arrays.equals` without any string conversion.